### ISSUE TITLE
ConsumerState

### ISSUE THREAD

**Issue by norbert-gaulia** [contek-io/io.contek.invoker#13]
RateLimitRule need to have option to set rate limited per hour or per minute, Bitmex websocket api allows 30 conncetions per hour, Rate limiter limits 20 per minute 
```
 RateLimiter createRateLimiter(String key, double cushion) {
    return RateLimiter.of(
        Joiner.on('_').join(type, name, key),
        RateLimiterConfig.custom()
            .limitForPeriod((int) (maxPermits * (1d - cushion)))
            .limitRefreshPeriod(Duration.of(1, MINUTES))
            .build());
  }
```

If IOException is common Okio keeps reconnecting more than 20 times per hour, which leads to 
429 Too Many Requests' following '403 Forbidden' following permaban of IP

I guess simply changing 
```
.limitRefreshPeriod(Duration.of(1, MINUTES))
```

with 
```
 .limitRefreshPeriod(resetPeriod)
```

Will do the trick?

**Comment by beijunyi** [comment]
I am not sure if I follow.

public static final RateLimitRule IP_WEB_SOCKET_CONNECTION_RULE =
        RateLimitRule.newBuilder()
            .setName("ip_web_socket_connection_rule")
            .setType(IP)
            .setMaxPermits(20)
            .setResetPeriod(Duration.ofHours(1))
            .build();

This code should limit WebSocket connection to 20 per hour. I think you are concerning that OkHttpClient's auto-retry feature may silently retry and break rate limit.

If this is the case, the proper solution shall be setting OkHttpClient's "retryOnConnectionFailure" field to false?

**Comment by norbertas-gaulia** [comment]
Please check this [line](https://github.com/contek-io/io.contek.invoker/blob/0710b6f75b82a9f488c95b2b53ee89cf54e8a166/commons/src/main/java/io/contek/invoker/commons/actor/ratelimit/RateLimitRule.java#L57) 

it doe not take into account provided resetPeriod, actually resetPeriod is not used anywhere.

**Comment by beijunyi** [comment]
You are right indeed. That was a terrible bug. Will fix and release 2.7.5-SNAPSHOT.
Your comment is much appreciated.

**Comment by norbertas-gaulia** [comment]
Also this [line ](https://github.com/contek-io/io.contek.invoker/blob/0710b6f75b82a9f488c95b2b53ee89cf54e8a166/commons/src/main/java/io/contek/invoker/commons/actor/ratelimit/RateLimitCache.java#L73)
 returns false if acquisition failed due to Timeout, default is 5 seconds, if acquire returns false it needs to bubble up (true/false result) to the caller and abandon the request if acquire returned false.

Right now this method relies on `Thread.park` of  `acquirePermission()`  method, but in 5 seconds it releases park and request is submitted every 5 seconds

**Comment by norbertas-gaulia** [comment]
I would be happy to add PR but i need you Intellij code formatting preset

**Comment by beijunyi** [comment]
I am using google-java-format plugin, which is free from Intellij plugin Market Place. It actually does more than the out-of-box IntelliJ code style. I also recommend "Save Action" plugin.

Regarding the new issue you pointed out. I will need to investigate a bit further before I can fix it. I will be very grateful if you can do a PR for this. Thanks!

**Comment by beijunyi** [comment]
Changing that line to RateLimiter.waitForPermission(limiter, permits);

Will this be enough?

**Comment by norbertas-gaulia** [comment]
Yep perfect, tested - does to job. Although to completely support Bitmex Bucket Re-Fill algorithm it needs a bit more play, but this is good enough. You guys have really some high quality lib here.

**Comment by beijunyi** [comment]
Awesome. The fix is included in 2.7.7-SNAPSHOT. Your contribution is much appreciated!
